module_3/gift_certificates_system/repository/src/main/java/com/epam/esm/dao/AbstractDao.java

```diff
@@ -0,0 +1,53 @@
+package com.epam.esm.dao;
+
+import org.springframework.beans.factory.annotation.Autowired;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.criteria.CriteriaBuilder;
+import java.util.Optional;
+
+/**
+ * Entity abstract DAO layer class.<br>
+ * Works with database.
+ *
+ * @param <T> Entity.
+ */
+public abstract class AbstractDao<T> extends FindAllDao<T> {
+    @PersistenceContext
+    protected EntityManager em;
+
+    @Autowired
+    protected CriteriaBuilder cb;
+
+    /**
+     * Create an entity in the database.
+     *
+     * @param entity - Entity being created.
+     * @return - Created entity.
+     */
+    public abstract T create(T entity);
```

Comment:
I suppose it can have implementation as well

Reply:
Thanks, fixed:
```java
public T create(T entity) {
        em.persist(entity);
        return entity;
    }
```